features/device-config/persistence/impl/src/main/java/org/opennms/features/deviceconfig/persistence/impl/DeviceConfigDaoImpl.java

```diff
@@ -82,6 +82,27 @@ public List<DeviceConfig> findConfigsForInterfaceSortedByDate(OnmsIpInterface ip
                 ipInterface.getId(), serviceName);
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public List<DeviceConfig> findStaleConfigs(OnmsIpInterface ipInterface, String serviceName,
+            Date staleDate, Optional<Long> excludedId) {
+        var builder = new CriteriaBuilder(DeviceConfig.class);
+        builder
+            .isNotNull("lastUpdated")
+            .isNotNull("config")
+            .eq("ipInterface.id", ipInterface.getId())
+            .eq("serviceName", serviceName)
+            .lt("lastUpdated", staleDate);
+
+        if (excludedId.isPresent()) {
```

Comment:
What if this is the first config that we are persisting and excludedId is not present then we don't need to make this query since we already know this is the only config for this interface and servicename.

Reply:
Good point.

In `DeviceConfigMonitorAdaptor.handlePollResult` we could check if `latestConfig.isPresent()`. If it is *not*, then we know right before `cleanupStaleConfigs` that the first `DeviceConfig` was added (in `deviceConfigDao.updateDeviceConfigContent`), and don't need to call `cleanupStaleConfigs`.

Reply:
Just added this, but need to test